server/src/main/java/org/elasticsearch/index/shard/IndexWriteLoadForecast.java

```diff
@@ -0,0 +1,162 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.index.shard;
+
+import org.elasticsearch.cluster.ClusterState;
+import org.elasticsearch.cluster.metadata.IndexAbstraction;
+import org.elasticsearch.cluster.metadata.IndexMetadata;
+import org.elasticsearch.cluster.metadata.Metadata;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.io.stream.Writeable;
+import org.elasticsearch.core.TimeValue;
+import org.elasticsearch.index.Index;
+import org.elasticsearch.index.IndexSettings;
+import org.elasticsearch.xcontent.ConstructingObjectParser;
+import org.elasticsearch.xcontent.ParseField;
+import org.elasticsearch.xcontent.ToXContentFragment;
+import org.elasticsearch.xcontent.XContentBuilder;
+import org.elasticsearch.xcontent.XContentParser;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.List;
+import java.util.OptionalDouble;
+import java.util.OptionalLong;
+
+public class IndexWriteLoadForecast implements Writeable, ToXContentFragment {
+    public static final ParseField SHARDS_WRITE_LOAD_FIELD = new ParseField("loads");
+
+    @SuppressWarnings("unchecked")
+    private static final ConstructingObjectParser<IndexWriteLoadForecast, Void> PARSER = new ConstructingObjectParser<>(
+        "index_write_load_forecast_parser",
+        false,
+        (args, unused) -> new IndexWriteLoadForecast(((List<Double>) args[0]).stream().mapToDouble(value -> value).toArray())
+    );
+
+    static {
+        PARSER.declareDoubleArray(ConstructingObjectParser.constructorArg(), SHARDS_WRITE_LOAD_FIELD);
+    }
+
+    private final double[] forecastedShardWriteLoad;
+
+    private IndexWriteLoadForecast(double[] forecastedShardWriteLoad) {
+        this.forecastedShardWriteLoad = forecastedShardWriteLoad;
+    }
+
+    public IndexWriteLoadForecast(StreamInput in) throws IOException {
+        this(in.readDoubleArray());
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        out.writeDoubleArray(forecastedShardWriteLoad);
+    }
+
+    public static ClusterState maybeIncludeWriteLoadForecast(
+        IndexAbstraction.DataStream dataStream,
+        ClusterState clusterState,
+        TimeValue maxIndexAge,
+        TimeValue minShardUptime
+    ) {
+        final IndexMetadata writeIndex = clusterState.metadata().getIndexSafe(dataStream.getWriteIndex());
+        if (IndexSettings.FORECAST_WRITE_LOAD_SETTING.get(writeIndex.getSettings()) == false) {
```

Comment:
This private setting is set by the new plugin in order to check if the cluster has the proper license to use this feature.